requests-for-comments/2023-11-22-scheduled-flag-changes.md

```diff
@@ -0,0 +1,121 @@
+            }],
+            "rollout_percentage": 20
+        }
+    }
+}
+```
+
+### Queue processing
+
+_At_ or _after_ the time specified, the change needs to be propagated onto the record.
+
+We already use Celery with a Redis queue to execute various background tasks. However, because the Redis queue is not durable, we cannot schedule tasks at the precise point in the future. A possible solution here is to have Postgres table of scheduled changes. Celery could periodically poll this table, e.g. every minute. During each poll, Celery would process any rows where the _scheduled_at_ time has already passed.
+
+The `scheduled_changes` table would have the following schema:
+
+| team_id | record_id | model       | changes (JSON) | scheduled_at (timestamp) | executed_at (timestamp, null) |
+| ------- | --------- | ----------- | -------------- | ------------------------ | ----------------------------- |
+| 1       | 57        | FeatureFlag | [ ... ]        | 2023-11-23 15:30:00      | null                          |
+
+The table would be indexed on the `scheduled_at` column to efficiently retrieve the columns that have the scheduled\*at value in the past. For every row retrieved, the task would **atomically**:
+
+-   Modify the field on the model record
+-   Mark the row as completed by setting the `executed_at` value
+
+#### Concurrency and retries
+
+To tackle the issue of concurrency and retries, let's consider the following questions.
+
+**What delay in mutations is acceptable?**
+
+-   A delay up to 1 minute should be acceptable; above that we might receive complaints. Example: users want to precisely time a feature release such as a ticket sale at noon.
+
+**Can mutations fail?**
+
+Mutations will fail in the following scenarios:
+
+-   The flag has been deleted in the meantime - no retry, just mark the row as completed.
+-   Some field is invalid (e.g. filters) - no retry, but record the reason for the failure. (To guard against this, we can validate the fields in the “schedule changes” API)
+-   PG is down - need to retry.
+
+Considering the above, our approach should be as follows:
+
+-   Retries: If PG downtime is the only reason for a failed mutation, we can simply wait for the next scheduled run of Celery at which point the schedule change will be retried.
+-   Concurrency: Given that our current/foreseeable count of records is low, we can afford concurrency=1. Even during peak loads, we should be able to handle all scheduled changes within 1 minute. To implement concurrency=1, we can use a transaction with the NOWAIT clause. The first worker that selects the rows "wins", while the queries from subsequent workers will fail with an error.
+
+### Model updates
+
+Every model that needs scheduled updates will implement a `scheduled_changes_dispatcher()` method. This method will receive the payload from the Celery task and update the record. After the successful update, the worker marks the row as completed by setting `executed_at`.
```

Comment:
```suggestion
Every model that needs scheduled updates will implement a `scheduled_changes_dispatcher()` method. This method will receive the payload from the Celery task and update the record. After the successful update, the worker marks the row as completed by setting `executed_at`.

Since this is sync and blocks the entire schedule queue, we can't have long running mutations with this system.
```